apps/builder/app/shared/sync/sync-stores.ts

```diff
@@ -192,25 +195,89 @@ const syncStoresState = (name: SyncEventSource, publish: Publish) => {
   };
 };
 
+const handshakeAndSyncStores = (
+  source: SyncEventSource,
+  publish: Publish,
+  sync: (publish: Publish) => () => void
+) => {
+  const actions: Parameters<typeof publish>[0][] = [];
+  let destinationReady = false;
+
+  // Until builder is ready store action in local cache
+  const publishAction: typeof publish = (action) => {
+    if (destinationReady) {
+      return publish(action);
+    }
+    actions.push(action);
+  };
+
+  const unsubscribe = sync(publishAction);
+
+  const handshakeAction = {
+    type: "handshake",
+    payload: {
+      source,
+    },
+  } as const;
+
+  publish(handshakeAction);
+  // Use setInterval as with SSR iframe can send handshake before builder is initialized and vice versa
+  const intervalHandle = setInterval(() => {
+    publish(handshakeAction);
+  }, 100);
+
+  const destinationStoreReady = subscribe("handshake", (payload) => {
+    if (source === payload.source) {
+      return;
+    }
+
+    if (destinationReady) {
+      return;
+    }
+
+    // We need to publish it here last time
+    publish(handshakeAction);
+
+    clearInterval(intervalHandle);
+
+    for (const action of actions) {
+      publish(action);
+    }
+
+    // cleanup
+    actions.length = 0;
```

Comment:
assign empty array?

Reply:
its a known trick for cleanin up arrays without creating new once, but could be a splice as easily lol